OneCommunityGlobal/HGNRest
# Shree create blue square reason scheduler forall

# Description
![bug_desc_1](https://github.com/OneCommunityGlobal/HGNRest/assets/41186831/3cf537e2-1d70-446c-b99a-dba1ea27ce42)

![bug_desc_2](https://github.com/OneCommunityGlobal/HGNRest/assets/41186831/c5ce0ee6-d55a-4960-890c-8d8559ca4dc5)


## Related PRS (if any):
To test this backend PR you need to checkout the [#1542](https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/pull/1542) frontend PR.

## Main changes explained:
-Updated src/controllers/reasonSchedulingController.js to remove the condition (scheduler available only to Owner and Admin role) and updated postReason and patchReason functions to generate an email to user role/you on clicking save button.

## How to test:
Tips to know before watching the youtube video given below.

1. As we are using Gmail API , make use of valid working gmail email id.
2. Sender email id (email id used to send emails) and receiver email id (email id which receives mail from the sender email id) can be the same as sender email id or different. Keep it the same to avoid any confusion.

Also, We need to understand the 4 different places where we need to make use of the valid working email id. 
1. In .env file (use sender email id).
2. Login on https://console.cloud.google.com/ as sender email id, so that it can verify the account that is used for sending the email.
3. Refer Video - The Test users step is missing in the video, to add test user - after creating scopes → Test users → add users (enter sender email id). Please do not miss this step.
![Test_Users](https://github.com/OneCommunityGlobal/HGNRest/assets/41186831/03b027d3-e2cb-4ef6-8cd6-8eaa1de3a112)
4. In HGN App → While updating a user role with valid working email id (can use the same sender email id or receiver email id - any valid working email id on which you want to receive the mail).

===========================================================================================

Before moving to testing, firstly you need to set up any role user with your own test email id, so that you will receive an email on this email id. This is only for test purposes.

1. To send email using Gmail API using OAuth2 and nodemailer, kindly follow the below video and set up a test mail id.
https://www.youtube.com/watch?v=-rcRf7yswfM&t=10s&ab_channel=MafiaCodes

2. Fill below variable values(fill inside single quotes) in the .env file of the backend.
REACT_APP_EMAIL =''
REACT_APP_EMAIL_CLIENT_ID =''
REACT_APP_EMAIL_CLIENT_SECRET =''
REACT_APP_EMAIL_REFRESH_TOKEN =''
REACT_APP_EMAIL_CLIENT_REDIRECT_URI='https://developers.google.com/oauthplayground'
sendEmail = TRUE

===========================================================================================
Steps how to test
1. check into current backend branch
2. do `npm install`, `npm run build` and `npm start` to run this PR locally
3. Login as Admin User → Update any user role with a valid/working email id using the User Management tab (so that you can receive an email on this email Id).
4. Note down this email used above.
5. Login as a updated user role (with valid email id) → go to dashboard → Welcome,user → view Profile → (below blue squares) Schedule Blue Square Reason → select a date and input a reason and click on save  (To save the reason - either add a new reason or edit an existing one) → this would generate an email on your valid user emailId 
6. Check if you received an email on the email id used for creating the new user role.

## Screenshots or videos of changes:

https://github.com/OneCommunityGlobal/HGNRest/assets/41186831/b1d5fa88-e005-4f14-8e64-d215925d7d4f

## Note:
The Refresh Token gets expired in 7 days, so if you are testing this PR after 7 days of creating the Refresh token then please create new OAuth 2.0 Client ID for the same project and update the variables (mentioned above) in .env file. (to create this follow video from 3.20 timestamp )
